File: go/indicators/common/pearsonscorrelationcoefficient/pearsonscorrelationcoefficient.go
package pearsonscorrelationcoefficient

//nolint: gofumpt
import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
)

// PearsonsCorrelationCoefficient computes Pearson's Correlation Coefficient (r) over a rolling window.
//
// Given two input series X and Y, it computes:
//
//	r = (n*sumXY - sumX*sumY) / sqrt((n*sumX2 - sumX^2) * (n*sumY2 - sumY^2))
//
// The indicator is not primed during the first length-1 updates.
type PearsonsCorrelationCoefficient struct {
	mu sync.RWMutex
	core.LineIndicator
	barFunc entities.BarFunc
	length  int
	windowX []float64
	windowY []float64
	count   int
	pos     int
	sumX    float64
	sumY    float64
	sumX2   float64
	sumY2   float64
	sumXY   float64
	primed  bool
}

// New returns an instance of the indicator created using supplied parameters.
func New(p *Params) (*PearsonsCorrelationCoefficient, error) {
	const (
		invalid = "invalid pearsons correlation coefficient parameters"
		fmts    = "%s: %s"
		fmtw    = "%s: %w"
		fmtn    = "correl(%d%s)"
		minlen  = 1
	)

	length := p.Length
	if length < minlen {
		return nil, fmt.Errorf(fmts, invalid, "length should be positive")
	}

	var (
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	bc := p.BarComponent
	qc := p.QuoteComponent
	tc := p.TradeComponent

	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	mnemonic := fmt.Sprintf(fmtn, length, core.ComponentTripleMnemonic(bc, qc, tc))
	desc := "Pearsons Correlation Coefficient " + mnemonic

	c := &PearsonsCorrelationCoefficient{
		barFunc: barFunc,
		length:  length,
		windowX: make([]float64, length),
		windowY: make([]float64, length),
	}

	c.LineIndicator = core.NewLineIndicator(mnemonic, desc, barFunc, quoteFunc, tradeFunc, c.Update)

	return c, nil
}

// IsPrimed indicates whether an indicator is primed.
func (s *PearsonsCorrelationCoefficient) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes an output data of the indicator.
func (s *PearsonsCorrelationCoefficient) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.PearsonsCorrelationCoefficient,
		s.LineIndicator.Mnemonic,
		s.LineIndicator.Description,
		[]core.OutputText{
			{Mnemonic: s.LineIndicator.Mnemonic, Description: s.LineIndicator.Description},
		},
	)
}

// Update updates the indicator given a single scalar sample.
// For a single-input update, both X and Y are set to the same value (degenerate case, always returns 1 or 0).
func (s *PearsonsCorrelationCoefficient) Update(sample float64) float64 {
	return s.UpdatePair(sample, sample)
}

// UpdatePair updates the indicator given an (x, y) pair.
func (s *PearsonsCorrelationCoefficient) UpdatePair(x, y float64) float64 {
	if math.IsNaN(x) || math.IsNaN(y) {
		return math.NaN()
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	n := float64(s.length)

	if s.primed {
		// Remove the oldest values.
		oldX := s.windowX[s.pos]
		oldY := s.windowY[s.pos]

		s.sumX -= oldX
		s.sumY -= oldY
		s.sumX2 -= oldX * oldX
		s.sumY2 -= oldY * oldY
		s.sumXY -= oldX * oldY

		// Add new values.
		s.windowX[s.pos] = x
		s.windowY[s.pos] = y
		s.pos = (s.pos + 1) % s.length

		s.sumX += x
		s.sumY += y
		s.sumX2 += x * x
		s.sumY2 += y * y
		s.sumXY += x * y

		return s.correlate(n)
	}

	// Accumulating phase.
	s.windowX[s.count] = x
	s.windowY[s.count] = y

	s.sumX += x
	s.sumY += y
	s.sumX2 += x * x
	s.sumY2 += y * y
	s.sumXY += x * y

	s.count++

	if s.count == s.length {
		s.primed = true
		s.pos = 0

		return s.correlate(n)
	}

	return math.NaN()
}

// correlate computes the Pearson correlation from the running sums.
func (s *PearsonsCorrelationCoefficient) correlate(n float64) float64 {
	varX := s.sumX2 - (s.sumX*s.sumX)/n
	varY := s.sumY2 - (s.sumY*s.sumY)/n
	tempReal := varX * varY

	if tempReal <= 0 {
		return 0
	}

	return (s.sumXY - (s.sumX*s.sumY)/n) / math.Sqrt(tempReal)
}

// UpdateBar updates the indicator given the next bar sample.
// This shadows LineIndicator.UpdateBar to extract both high (X) and low (Y) from the bar.
func (s *PearsonsCorrelationCoefficient) UpdateBar(sample *entities.Bar) core.Output {
	x := sample.High
	y := sample.Low
	value := s.UpdatePair(x, y)

	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: value}

	return output
}

File: py/indicators/common/pearsons_correlation_coefficient/pearsons_correlation_coefficient.py
"""Pearson's correlation coefficient indicator."""

import math

from ...core.line_indicator import LineIndicator
from ...core.metadata import Metadata
from ...core.build_metadata import build_metadata, OutputText
from ...core.identifier import Identifier
from ...core.component_triple_mnemonic import component_triple_mnemonic
from ...core.output import Output
from ....entities.bar import Bar
from ....entities.scalar import Scalar
from ....entities.bar_component import BarComponent, DEFAULT_BAR_COMPONENT, bar_component_value
from ....entities.quote_component import QuoteComponent, DEFAULT_QUOTE_COMPONENT, quote_component_value
from ....entities.trade_component import TradeComponent, DEFAULT_TRADE_COMPONENT, trade_component_value
from .params import PearsonsCorrelationCoefficientParams


class PearsonsCorrelationCoefficient(LineIndicator):
    """Computes Pearson's Correlation Coefficient (r) over a rolling window.

    Given two input series X and Y, it computes:
        r = (n*sumXY - sumX*sumY) / sqrt((n*sumX2 - sumX^2) * (n*sumY2 - sumY^2))

    For single-input updates, X=Y=sample (degenerate case).
    For bar updates, X=High, Y=Low.
    """

    def __init__(self, params: PearsonsCorrelationCoefficientParams) -> None:
        length = params.length
        if length < 1:
            raise ValueError(
                "invalid pearsons correlation coefficient parameters: length should be positive")

        bc = params.bar_component if params.bar_component is not None else DEFAULT_BAR_COMPONENT
        qc = params.quote_component if params.quote_component is not None else DEFAULT_QUOTE_COMPONENT
        tc = params.trade_component if params.trade_component is not None else DEFAULT_TRADE_COMPONENT

        bar_func = bar_component_value(bc)
        quote_func = quote_component_value(qc)
        trade_func = trade_component_value(tc)

        mnemonic = f"correl({length}{component_triple_mnemonic(bc, qc, tc)})"
        description = f"Pearsons Correlation Coefficient {mnemonic}"

        super().__init__(mnemonic, description, bar_func, quote_func, trade_func, self.update)

        self._bar_func_raw = bar_func
        self._length = length
        self._window_x: list[float] = [0.0] * length
        self._window_y: list[float] = [0.0] * length
        self._count: int = 0
        self._pos: int = 0
        self._sum_x: float = 0.0
        self._sum_y: float = 0.0
        self._sum_x2: float = 0.0
        self._sum_y2: float = 0.0
        self._sum_xy: float = 0.0
        self._primed: bool = False

    def is_primed(self) -> bool:
        return self._primed

    def metadata(self) -> Metadata:
        return build_metadata(
            Identifier.PEARSONS_CORRELATION_COEFFICIENT,
            self.mnemonic,
            self.description,
            [OutputText(self.mnemonic, self.description)],
        )

    def update(self, sample: float) -> float:
        """Single-input update: sets X=Y=sample."""
        return self.update_pair(sample, sample)

    def update_pair(self, x: float, y: float) -> float:
        """Update with an (x, y) pair."""
        if math.isnan(x) or math.isnan(y):
            return math.nan

        n = float(self._length)

        if self._primed:
            old_x = self._window_x[self._pos]
            old_y = self._window_y[self._pos]

            self._sum_x -= old_x
            self._sum_y -= old_y
            self._sum_x2 -= old_x * old_x
            self._sum_y2 -= old_y * old_y
            self._sum_xy -= old_x * old_y

            self._window_x[self._pos] = x
            self._window_y[self._pos] = y
            self._pos = (self._pos + 1) % self._length

            self._sum_x += x
            self._sum_y += y
            self._sum_x2 += x * x
            self._sum_y2 += y * y
            self._sum_xy += x * y

            return self._correlate(n)

        # Accumulating phase.
        self._window_x[self._count] = x
        self._window_y[self._count] = y

        self._sum_x += x
        self._sum_y += y
        self._sum_x2 += x * x
        self._sum_y2 += y * y
        self._sum_xy += x * y

        self._count += 1

        if self._count == self._length:
            self._primed = True
            self._pos = 0
            return self._correlate(n)

        return math.nan

    def _correlate(self, n: float) -> float:
        var_x = self._sum_x2 - (self._sum_x * self._sum_x) / n
        var_y = self._sum_y2 - (self._sum_y * self._sum_y) / n
        temp_real = var_x * var_y

        if temp_real <= 0:
            return 0.0

        return (self._sum_xy - (self._sum_x * self._sum_y) / n) / math.sqrt(temp_real)

    def update_bar(self, sample: Bar) -> Output:
        """Custom bar update: X=High, Y=Low."""
        x = sample.high
        y = sample.low
        value = self.update_pair(x, y)
        return [Scalar(sample.time, value)]
